pkg/vsphere/performance/host_metrics.go

```diff
@@ -0,0 +1,221 @@
+	}
+
+	morefToHost := make(map[string]*object.HostSystem)
+	morefs := make([]types.ManagedObjectReference, len(hosts))
+	for i, host := range hosts {
+		moref := host.Reference()
+		morefToHost[moref.String()] = host
+		morefs[i] = moref
+	}
+
+	// Query CPU and memory metrics for the morefs.
+	spec := types.PerfQuerySpec{
+		Format:     string(types.PerfFormatNormal),
+		IntervalId: sampleInterval,
+	}
+
+	counters := []string{cpuUsage, memActive, memConsumed, memTotalCapacity, memOverhead}
+	perfMgr := performance.NewManager(h.Client)
+	sample, err := perfMgr.SampleByName(op.Context, spec, counters, morefs)
+	if err != nil {
+		errStr := "unable to get metric sample: %s"
+		op.Errorf(errStr, err)
+		return nil, fmt.Errorf(errStr, err)
+	}
+
+	results, err := perfMgr.ToMetricSeries(op.Context, sample)
+	if err != nil {
+		errStr := "unable to convert metric sample to metric series: %s"
+		op.Errorf(errStr, err)
+		return nil, fmt.Errorf(errStr, err)
+	}
+
+	metrics := assembleMetrics(op, morefToHost, results)
+	return metrics, nil
+}
+
+// assembleMetrics processes the metric samples received from govmomi and returns a finalized metrics map
+// keyed by the hosts.
+func assembleMetrics(op trace.Operation, morefToHost map[string]*object.HostSystem,
+	results []performance.EntityMetric) map[string]*HostMetricsInfo {
+	metrics := make(map[string]*HostMetricsInfo)
+
+	for _, host := range morefToHost {
+		metrics[host.Reference().String()] = &HostMetricsInfo{}
+	}
+
+	for i := range results {
+		res := results[i]
+		host, exists := morefToHost[res.Entity.String()]
+		if !exists {
+			op.Warnf("moref %s does not exist in requested morefs, skipping", res.Entity.String())
+			continue
+		}
+
+		ref := host.Reference().String()
+		// Process each value and assign it directly to the corresponding metric field
+		// since there is only one sample.
+		for _, v := range res.Value {
+
+			// We don't need to collect non-aggregate (non-empty Instance) metrics.
+			if v.Instance != "" {
+				continue
+			}
+
+			if len(v.Value) == 0 {
+				op.Warnf("metric %s for moref %s has no value, skipping", v.Name, res.Entity.String())
+				continue
+			}
+
+			switch v.Name {
+			case cpuUsage:
+				// Convert percent units from 1/100th of a percent (100 = 1%) to a human-readable percentage.
+				metrics[ref].CPU.UsagePercent = float64(v.Value[0]) / 100.0
+			case memActive:
+				metrics[ref].Memory.ActiveKB = v.Value[0]
+			case memConsumed:
+				metrics[ref].Memory.ConsumedKB = v.Value[0]
+			case memOverhead:
+				metrics[ref].Memory.OverheadKB = v.Value[0]
+			case memTotalCapacity:
+				// Total capacity is in MB, convert to KB so as to have all memory values in KB.
+				metrics[ref].Memory.TotalKB = v.Value[0] * 1024
+			}
+		}
+	}
+
+	return metrics
+}
+
+// filterHosts removes candidate hosts who are either disconnected or in maintenance mode.
+func filterHosts(op trace.Operation, s *session.Session, hosts []*object.HostSystem) ([]*object.HostSystem, error) {
```

Comment:
To my knowledge you can't relocate a VM to a host that is disconnected or in maintenance mode as the relocate compat check would fail. That being said, attempting to gather metrics on a host that is in one of these states might cause problems. For that reason, I'm adding a call to this func before we actually reach out to gather host metrics. I think it was intended to be there all along, but got lost in the churn.

Reply:
https://github.com/vmware/vic/pull/7704